# Programming Assignment For Graphs

The dataset chosed for the implementation is Wikepedia dataset
Dataset Link:  https://snap.stanford.edu/data/wiki-Vote.html

The Libraries used for the program are networkx ,community.

In [38]:
import networkx as nx
import community as community_louvain
from collections import defaultdict


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Loading Wikipedia Dataset using edgelist method

In [39]:
GW=nx.read_edgelist("/content/drive/MyDrive/wiki-Vote.txt", create_using=nx.DiGraph())

nx.info(GW)

'Name: \nType: DiGraph\nNumber of nodes: 7115\nNumber of edges: 103689\nAverage in degree:  14.5733\nAverage out degree:  14.5733'

Loading Wikepedia dataset using read_edgelist method, As Wikepedia dataset is directed to initilaze the graph i am using DiGraph()

# 1. Finding Number Of Nodes

In [42]:
no_nodes= nx.number_of_nodes(GW)
print('Number of Nodes: ',no_nodes)

Number of Nodes:  7115


Using number_of_nodes() method found number of nodes in the graph are - 7115

# 2. Finding Number Of Edges

In [43]:
no_edges = nx.number_of_edges(GW)
print('Number of Edges: ',no_edges)

Number of Edges:  103689


Using number_of_edges() method found number of edges in graph are - 103689

# 3. Checking Graph strongly connected or not?

In [45]:
nx.is_strongly_connected(GW)


False

From the results we can conclude the Graph is not strongly connected.

# Checking for number of strongly and weakly connected components?

In [50]:
weakly_components= nx.number_weakly_connected_components(GW)
print('weakly_components',weakly_components) 

strongly_components = nx.number_strongly_connected_components(GW)
print('strongly_components',strongly_components)

weakly_components 24
strongly_components 5816


After applying the respective methods for finding strong and weakly connected components found 24 weakly connected and 5816 strongly components

# 4. Diameter of the Graph

In [ ]:
print(nx.diameter(GW))

Diameter for not strongly connected graph is diameter is infinite.
While running this code block it throws an error mentioning same.

# 5. Finding the average clustering coefficient

In [49]:
avgclustering_coef = nx.average_clustering(GW)
print(avgclustering_coef)

0.08156344522820935


average clustering coefficient for garph is 0.0815

# 6a) Finding  two nodes with highest closness centarlity

In [24]:
cc = nx.closeness_centrality(GW)
lst = sorted(cc, key=cc.get, reverse=True)[:2]
for i in lst:
    print("Node:",i)


Node: 4037
Node: 15


After finding closness centarlity for all the nodes in the graph. To get top 2 nodes with highest closness centrality, the return value type is a dictionary with nodes and there respective centrality so using sorted method and getting the top two nodes with highest centraily.After the process top 2 nodes with highest closness centrality are NODE 4037 and 15

# 6b)Finding  two nodes with highest betweenness centarlity 

In [44]:
bc = nx.betweenness_centrality(GW)
lst = sorted(bc, key=bc.get, reverse=True)[:2]
for i in lst:
    print("Node:",i)


Node: 2565
Node: 1549


After finding betweenness centarlity for all the nodes in the graph. To get top 2 nodes with highest betweenness centrality the return value type is a dictionary with nodes and there respective centrality so using sorted method and getting the top two nodes with highest centraily.After the process top 2 nodes with highest betweenness centrality are NODE 2565 and 1549.

# 6c)Finding two nodes with highest Page Rank 

In [48]:
pr = nx.pagerank(GW)
lst = sorted(pr, key=pr.get, reverse=True)[:2]
for i in lst:
    print("Node:",i)


Node: 4037
Node: 15


To find top 2 nodes with highest page rank using pagerank() method from nx library.The return value type is a dictionary with nodes and there respective pagerank so using sorted method and getting the top two nodes with highest page rank.After the process top 2 nodes with highest pagerank are NODE 4037 and 15

# 7.Dispersion between two nodes with high page rank

In [47]:
dispersion = nx.dispersion(GW,u="4037",v="15")
print(dispersion)


0.0


Using nx.dispersion() and passing two top nodes with high pagerank dispersion value is 0.0

# 8a)Similarity between two nodes with highest betweenness

For finding similarity rank between two nodes with highest betweenness and page rank,Used simrank_similarity() method from networkx library. Considered maximum iterations =1 due to time constraint.

In [ ]:
sim= nx.simrank_similarity(GW,'2565','1549',max_iterations=1)


In [10]:
print('The simrank similarity between the two nodes with highest betweenness centrality: ',sim)

simrank similarity between the two nodes with highest betweenness centrality:  0.00116639356472516


# b)SimRank similarity between the two nodes with highest PageRank

In [11]:
simrank_pagerank= nx.simrank_similarity(GW,'4037','15',max_iterations=1)

In [12]:
print('The SimRank similarity between the two nodes with highest PageRank: ',simrank_pagerank)

The SimRank similarity between the two nodes with highest PageRank:  0.000578262424459167


**9) Finding communities using Louvain algorithm**

As louvain algorithm considers only undirected graph, converting directed graph to undirected using to_undirected() method.
Using community_louvain.best_partition method finding communities between graph

In [27]:
G = GW.to_undirected()
communities = community_louvain.best_partition(G)



Printing the all the communites found

In [28]:
communities

{'30': 6,
 '1412': 5,
 '3352': 25,
 '5254': 3,
 '5543': 3,
 '7478': 7,
 '3': 6,
 '28': 6,
 '39': 6,
 '54': 6,
 '108': 6,
 '152': 6,
 '178': 6,
 '182': 6,
 '214': 6,
 '271': 6,
 '286': 6,
 '300': 6,
 '348': 6,
 '349': 6,
 '371': 6,
 '567': 6,
 '581': 6,
 '584': 6,
 '586': 6,
 '590': 6,
 '604': 6,
 '611': 6,
 '8283': 6,
 '25': 6,
 '6': 6,
 '8': 6,
 '19': 6,
 '23': 6,
 '29': 6,
 '33': 6,
 '35': 6,
 '50': 6,
 '55': 6,
 '75': 6,
 '80': 6,
 '86': 6,
 '94': 6,
 '127': 6,
 '154': 6,
 '192': 6,
 '224': 6,
 '226': 6,
 '246': 6,
 '252': 6,
 '255': 6,
 '257': 6,
 '261': 6,
 '273': 6,
 '280': 6,
 '282': 6,
 '299': 6,
 '302': 6,
 '306': 6,
 '308': 6,
 '310': 6,
 '322': 6,
 '325': 6,
 '339': 6,
 '356': 6,
 '358': 6,
 '363': 6,
 '370': 6,
 '373': 6,
 '377': 6,
 '379': 6,
 '407': 5,
 '415': 6,
 '422': 6,
 '428': 6,
 '439': 6,
 '488': 6,
 '514': 6,
 '545': 6,
 '549': 6,
 '560': 6,
 '592': 6,
 '600': 6,
 '645': 6,
 '650': 6,
 '659': 6,
 '664': 6,
 '667': 6,
 '673': 6,
 '675': 6,
 '696': 6,
 '706': 6,
 '7

Grouping the communities based on thier respective values in dictionary

In [37]:
result = defaultdict(list)
for key, val in sorted(communities.items()):
    result[val].append(key)
     
print("Grouped communities are : " + str(dict(result)))

Grouped communities are : {6: ['10', '100', '1001', '1002', '1007', '1008', '101', '1013', '1014', '102', '1020', '1021', '1022', '1023', '1025', '1026', '1027', '1028', '1029', '103', '1030', '1033', '1038', '1039', '104', '1040', '1043', '1044', '1045', '1046', '1047', '1048', '105', '1052', '1053', '106', '1064', '1066', '1067', '107', '1070', '1071', '1072', '1075', '1077', '1078', '1079', '108', '1080', '1081', '1083', '1084', '1085', '1087', '1088', '1090', '1092', '1093', '1094', '1096', '110', '1101', '1104', '1105', '1106', '1107', '1108', '1109', '111', '1110', '1112', '1113', '1114', '1115', '112', '1120', '1123', '1125', '1126', '1128', '113', '1130', '1131', '1134', '1138', '1139', '114', '1140', '1141', '1144', '1145', '1146', '115', '1152', '1155', '1156', '1158', '116', '1160', '1161', '117', '1170', '118', '1187', '1188', '1189', '119', '1190', '1192', '1193', '1194', '1195', '1196', '12', '120', '1204', '121', '1218', '1219', '1220', '1222', '123', '1232', '1235', '12